tests/test_docs.py

```diff
@@ -0,0 +1,449 @@
+F = pytest.importorskip("torch.autograd.functional")
+jax = pytest.importorskip("jax")
+jnp = pytest.importorskip("jax.numpy")
+
+devices = ["default.qubit"]
+interfaces = [None, "autograd", "jax", "tf", "torch"]
+diff_interfaces = ["autograd", "jax", "tf", "torch"]
+shotss = [None, 100]
+
+diff_methods = ["device", "backprop", "adjoint", "parameter-shift", "finite-diff"]
+return_types = [
+    State,
+    "DensityMatrix",
+    Probability,
+    Sample,
+    Expectation,
+    "Hermitian",
+    "Projector",
+    Variance,
+]
+
+grad_types = [Expectation, "Hermitian", "Projector", Variance]
+jacob_types = [State, "DensityMatrix", Probability, Sample]
+
+
+def get_qnode(interface, diff_method, return_type, shots):
+    """Return a QNode with the given attributes"""
+    dev = qml.device("default.qubit", wires=1, shots=shots)
+
+    @qml.qnode(dev, interface=interface, diff_method=diff_method)
+    def circuit(x):
+        qml.Hadamard(wires=0)
+        qml.RX(x[0], wires=0)
+        if return_type == State:
+            return qml.state()
+        elif return_type == "DensityMatrix":
+            return qml.density_matrix(wires=0)
+        elif return_type == Probability:
+            return qml.probs(wires=0)
+        elif return_type == Sample:
+            return qml.sample(wires=0)
+        elif return_type == Expectation:
+            return qml.expval(qml.PauliZ(wires=0))
+        elif return_type == "Hermitian":
+            return qml.expval(qml.Hermitian(np.array([[1.0, 0.0], [0.0, -1.0]]), wires=0))
+        elif return_type == "Projector":
+            return qml.expval(qml.Projector(np.array([1]), wires=0))
+        elif return_type == Variance:
+            return qml.var(qml.PauliZ(wires=0))
+
+    return circuit
+
+
+def get_variable(interface):
+    """Return an interface-specific trainable variable"""
+    if interface is None:
+        return np.array([0.1], requires_grad=True)
+    elif interface == "autograd":
+        return np.array([0.1], requires_grad=True)
+    elif interface == "jax":
+        return jnp.array([0.1])
+    elif interface == "tf":
+        return tf.Variable([0.1], trainable=True)
+    elif interface == "torch":
+        return torch.tensor([0.1], requires_grad=True)
+
+
+def compute_gradient(x, interface, circuit, return_type):
+    """Return an interface-specific gradient or jacobian"""
+    if interface == "autograd":
+        if return_type in grad_types:
+            return qml.grad(circuit)(x)
+        elif return_type == State:
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = np.abs(res) ** 2
+                return probs[0]
+
+            return qml.grad(cost_fn)(x)
+        elif return_type == "DensityMatrix":
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = np.abs(res) ** 2
+                return probs[0][0]
+
+            return qml.grad(cost_fn)(x)
+        else:
+            return qml.jacobian(circuit)(x)
```

Comment:
> Right now computing the gradient of qml.sample doesn't raise an error.

Oh really? What is the behaviour here? :thinking: That's really odd.

Reply:
The gradient computation is allowed with either `parameter-shift` or `finite-diff`, but the gradient itself looks incorrect. This behaviour is unintuitive since taking the gradient of a sample doesn't really make any sense.

Ideally we still want to support the forward pass when the return value of a QNode is `qml.sample()`, but fail if the user attempts to compute the gradient.

Reply:
Created an issue [here](https://github.com/PennyLaneAI/pennylane/issues/2556).

Reply:
Leaving the test case as-is, since we'll update the docs once we fix the issue.